pageserver/src/walingest.rs

```diff
@@ -151,56 +151,50 @@ impl<'a> WalIngest<'a> {
+                        // So we can reuse XlCreateDatabase here.
+                        debug!("XLOG_DBASE_CREATE_WAL_LOG");
+                        let createdb = XlCreateDatabase::decode(&mut buf);
+                        self.ingest_xlog_dbase_create(modification, &createdb, ctx)
                             .await?;
                     }
+                    postgres_ffi::v15::bindings::XLOG_DBASE_DROP => {
+                        let dropdb = XlDropDatabase::decode(&mut buf);
+                        for tablespace_id in dropdb.tablespace_ids {
+                            trace!("Drop db {}, {}", tablespace_id, dropdb.db_id);
+                            modification
+                                .drop_dbdir(tablespace_id, dropdb.db_id, ctx)
+                                .await?;
+                        }
+                    }
+                    _ => {}
                 }
             } else if self.timeline.pg_version == 16 {
-                if (decoded.xl_info & pg_constants::XLR_RMGR_INFO_MASK)
-                    == postgres_ffi::v16::bindings::XLOG_DBASE_CREATE_WAL_LOG
-                {
-                    debug!("XLOG_DBASE_CREATE_WAL_LOG: noop");
-                } else if (decoded.xl_info & pg_constants::XLR_RMGR_INFO_MASK)
-                    == postgres_ffi::v16::bindings::XLOG_DBASE_CREATE_FILE_COPY
-                {
-                    // The XLOG record was renamed between v14 and v15,
-                    // but the record format is the same.
-                    // So we can reuse XlCreateDatabase here.
-                    debug!("XLOG_DBASE_CREATE_FILE_COPY");
-                    let createdb = XlCreateDatabase::decode(&mut buf);
-                    self.ingest_xlog_dbase_create(modification, &createdb, ctx)
-                        .await?;
-                } else if (decoded.xl_info & pg_constants::XLR_RMGR_INFO_MASK)
-                    == postgres_ffi::v16::bindings::XLOG_DBASE_DROP
-                {
-                    let dropdb = XlDropDatabase::decode(&mut buf);
-                    for tablespace_id in dropdb.tablespace_ids {
-                        trace!("Drop db {}, {}", tablespace_id, dropdb.db_id);
-                        modification
-                            .drop_dbdir(tablespace_id, dropdb.db_id, ctx)
+                match decoded.xl_info & pg_constants::XLR_RMGR_INFO_MASK {
+                    postgres_ffi::v16::bindings::XLOG_DBASE_CREATE_WAL_LOG
+                    | postgres_ffi::v16::bindings::XLOG_DBASE_CREATE_FILE_COPY => {
+                        // The XLOG record was renamed between v14 and v15,
+                        // but the record format is the same.
+                        // So we can reuse XlCreateDatabase here.
+                        debug!("XLOG_DBASE_CREATE_WAL_LOG");
+                        let createdb = XlCreateDatabase::decode(&mut buf);
+                        self.ingest_xlog_dbase_create(modification, &createdb, ctx)
```

Comment:
Sorry, I do not understand you.
When we execute `create database tested`; then WAL record `XLOG_DBASE_CREATE_WAL_LOG` is created.

Reply:
Ok, now I understand your point.
And you are right - we are doing redundant job in  `ingest_xlog_dbase_create`.
But we do nothing in `ingest_xlog_dbase_create`, then there will be race condition and we really can get "key not found" error.

Please notice, that the first thing `CreateDatabaseUsingWalLog` is doing - it calls `CreateDirAndVersionFile` which in turn first of all write `XLOG_DBASE_CREATE_WAL_LOG` record and calls  XLOG_DBASE_CREATE_WAL_LOG. But non database files are copied at this moment. So if after creation of the database we execute some command which calls `smgrexists(...,FSM_FORKNUM)` then as request LSN it will get LSN of `XLOG_DBASE_CREATE_WAL_LOG`.
But since handler of this command in walingest does nothing, we will not find correspondent rel_dir object in our key value storage which cause "key not found" error.

How it can bee fixed?
One straightforward solution I see at this moment is to update last written LSN after `CreateDatabaseUsingWalLog`:

```
		if (dbstrategy == CREATEDB_WAL_LOG)
			CreateDatabaseUsingWalLog(src_dboid, dboid, src_deftablespace,
									  dst_deftablespace);
		else
			CreateDatabaseUsingFileCopy(src_dboid, dboid, src_deftablespace,
										dst_deftablespace);
+++	       SetLastWrittenLSNForDatabase(XactLastRecEnd);
```

Do you have some better proposal?

Reply:
Do we actually need to perform some actions for `XLOG_DBASE_CREATE_WAL_LOG` in walingest?
I looked at `ingest_xlog_dbase_create` and didn't find something which has to be done.
The fact that `create database` is working now when  `ingest_xlog_dbase_create` is not called also proves it.

So should I just remove this PR and add new one (actually 4 - for each PG version) which adds `SetLastWrittenLSNForDatabase(XactLastRecEnd);` after CreateDatabase* ?
Changing Postgres core is really awful now:(